API_GUIDE.md

```diff
@@ -1,76 +1,118 @@
+print an XLA tensor:
 
 ```python
 import torch
 import torch_xla
 import torch_xla.core.xla_model as xm
 
-x = torch.randn(4, 2, device=xm.xla_device())
-print(x.device)
-print(x)
+t = torch.randn(2, 2, device=xm.xla_device())
+print(t.device)
+print(t)
+```
+
+This code should look familiar. PyTorch/XLA uses the same interface as regular
+PyTorch with a few additions. Importing `torch_xla` initializes PyTorch/XLA, and
+`xm.xla_device()` returns the current XLA device. This may be a CPU or Cloud TPU
+depending on your environment.
+
+## XLA Tensors are PyTorch Tensors
+
+PyTorch operations can be performed on XLA tensors just like CPU or CUDA tensors.
+
+For example, XLA tensors can be added together:
+
+```python
+t0 = torch.randn(2, 2, device=xm.xla_device())
+t1 = torch.randn(2, 2, device=xm.xla_device())
+print(t0 + t1)
+```
+
+Or matrix multiplied:
+
+```python
+print(t0.mm(t1))
+```
+
+Or used with neural network modules:
+
+```python
+l_in = torch.randn(10, device=xm.xla_device())
+linear = torch.nn.Linear(10, 20).to(xm.xla_device())
+l_out = linear(l_in)
+print(l_out)
 ```
 
-The XLA device is not a physical device but instead stands in for either a Cloud TPU or CPU. The underlying storage for XLA tensors is a contiguous buffer in device memory and the code in the model shouldn't assume any stride.
+Like other device types, XLA tensors only work with other XLA tensors on the
+same device. So code like
 
-XLA Tensor doesn't support converting single tensor to half precision using `tensor.half()`. Instead, environment variable `XLA_USE_BF16` is available, which converts **all** PyTorch float values to bfloat16 when sending them to the TPU device. The conversion is totally transparent to the user, and the XLA tensors will still retain a float dtype. Similarly, when the tensor is moved back to CPU, its type will be float.
+```python
+l_in = torch.randn(10, device=xm.xla_device())
+linear = torch.nn.Linear(10, 20).cpu()
+l_out = linear(l_in)
+print(l_out)
+# Input tensor is not an XLA tensor: torch.FloatTensor
+```
 
-The [XLA readme](https://github.com/pytorch/xla/blob/master/README.md) describes all the options available to run on TPU or CPU.
+will throw an error since the torch.nn.Linear module is on the CPU.
 
-## Running a model
+## Running Models on XLA Devices
 
-There are different ways to run a model using the PyTorch/XLA framework.
+Building a new PyTorch network or converting an existing one to run on XLA devices requires only a few lines of XLA-specific code. The following snippets highlight these lines when running on a single device, multiple devices with XLA multiprocessing, or multiple threads with XLA multithreading.
 
-### Native PyTorch API
+### Running on a Single XLA Device
 
-The simplest (but not good performing) one is to just run on one core and send the input tensors to the XLA devices manually:
+The following snippet shows a network training on one of several
+device types:
```

Comment:
Some lines are really long, some instead seem to wrap around even though they'd fit an 80 column.
Would be nice to decide a standard and stick to it.
IMHO an 80 columns break look nicer on text editors.

Reply:
Good catch. Mostly under 80 columns now.

Reply:
Right.